File: config/ports.ts
import { existsSync, readFileSync } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const configDir = dirname(fileURLToPath(import.meta.url))

export interface Ports {
  serverPort: number
  clientPort: number
  clientHost: string
  clientOrigin: string
  serverOrigin: string
}

function parsePortsEnv(text: string): Record<string, string> {
  const values: Record<string, string> = {}
  for (const line of text.split('\n')) {
    const trimmed = line.trim()
    if (!trimmed || trimmed.startsWith('#')) continue
    const eq = trimmed.indexOf('=')
    if (eq === -1) continue
    values[trimmed.slice(0, eq).trim()] = trimmed.slice(eq + 1).trim()
  }
  return values
}

export function loadPorts(repoRoot = resolve(configDir, '..')): Ports {
  const path = resolve(repoRoot, 'config/ports.env')
  const fileValues = existsSync(path) ? parsePortsEnv(readFileSync(path, 'utf8')) : {}

  const serverPort = Number(
    process.env.PHOTOFRAME_SERVER_PORT ?? fileValues.PHOTOFRAME_SERVER_PORT,
  )
  const clientPort = Number(
    process.env.PHOTOFRAME_CLIENT_PORT ?? fileValues.PHOTOFRAME_CLIENT_PORT,
  )
  const clientHost =
    process.env.PHOTOFRAME_CLIENT_HOST ?? fileValues.PHOTOFRAME_CLIENT_HOST ?? 'localhost'

  if (!Number.isFinite(serverPort) || !Number.isFinite(clientPort)) {
    throw new Error(`Invalid or missing ports in ${path}`)
  }

  return {
    serverPort,
    clientPort,
    clientHost,
    clientOrigin: `http://${clientHost}:${clientPort}`,
    serverOrigin: `http://${clientHost}:${serverPort}`,
  }
}

File: server/app/ports.py
"""Load canonical port defaults from config/ports.env (overridable via environment)."""

from __future__ import annotations

import os
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path


@dataclass(frozen=True, slots=True)
class PortDefaults:
    server_port: int
    client_port: int
    client_host: str

    @property
    def client_origin(self) -> str:
        return f"http://{self.client_host}:{self.client_port}"

    @property
    def server_origin(self) -> str:
        return f"http://{self.client_host}:{self.server_port}"


def repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def ports_env_path() -> Path:
    return repo_root() / "config" / "ports.env"


def _parse_ports_env(text: str) -> dict[str, str]:
    values: dict[str, str] = {}
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        key, sep, value = stripped.partition("=")
        if sep:
            values[key.strip()] = value.strip()
    return values


def _require_int(name: str, file_values: dict[str, str]) -> int:
    if name in os.environ:
        return int(os.environ[name])
    if name in file_values:
        return int(file_values[name])
    msg = f"Missing required port setting {name} (expected in config/ports.env or the environment)"
    raise ValueError(msg)


@lru_cache(maxsize=1)
def load_port_defaults() -> PortDefaults:
    path = ports_env_path()
    if not path.is_file():
        msg = f"Missing {path}"
        raise FileNotFoundError(msg)

    file_values = _parse_ports_env(path.read_text(encoding="utf-8"))
    client_host = os.environ.get("PHOTOFRAME_CLIENT_HOST") or file_values.get(
        "PHOTOFRAME_CLIENT_HOST", "localhost"
    )

    return PortDefaults(
        server_port=_require_int("PHOTOFRAME_SERVER_PORT", file_values),
        client_port=_require_int("PHOTOFRAME_CLIENT_PORT", file_values),
        client_host=client_host,
    )
